src/algorithms/guidedpf.jl

```diff
@@ -0,0 +1,164 @@
+export GuidedFilter, GPF, AbstractProposal
+
+## PROPOSALS ###############################################################################
+"""
+    AbstractProposal
+"""
+abstract type AbstractProposal end
+
+function SSMProblems.distribution(
+    model::AbstractStateSpaceModel,
+    prop::AbstractProposal,
```

Comment:
I don't think we should change the `SSMProblems` interface. The proposal should be part of the `filter` interface, maybe something along the lines of:

```julia
abstract type AbstractProposal end

abstract type AbstractParticleFilter{N, P<:AbstractProposal} end 

struct ParticleFilter{N,RS,P} <: AbstractParticleFilter{N,P}
    resampler::RS
    proposal::P
end

# Default to latent dynamics
struct LatentProposal <: AbstractProposal end

const BootstrapFilter{N,RS} = ParticleFilter{N,RS,LatentProposal}
const BF = BootstrapFilter

function propose(
    rng::AbstractRNG, 
    prop::LatentProposal, 
    model::AbstractStateSpaceModel, 
    particles::ParticleContainer, 
    step, 
    state, 
    obs; 
    kwargs...
)
    return SSMProblems.simulate(rng, model.dyn, t, state; kwargs...)
end

function logdensity(prop::AbstractProposal, ...)
   return SSMProblmes.logdensity(...)
end
```

And we should probably update the `filter/predict` functions:
```julia
function predict(
    rng::AbstractRNG,
    model::StateSpaceModel,
    filter::BootstrapFilter,
    step::Integer,
    states::ParticleContainer{T};
    ref_state::Union{Nothing,AbstractVector{T}}=nothing,
    kwargs...,
) where {T}
    states.proposed, states.ancestors = resample(
        rng, filter.resampler, states.filtered, filter
    )
    states.proposed.particles = map(states.proposed) do state
        propose(rng, filter.proposal, model.dyn, step, state; kwargs...),
    end

    return update_ref!(states, ref_state, filter, step)
end
```

Reply:
I 100% agree with the BF integration, I was intentionally working my way up to that, but didn't want to drastically change the interface upon the first commit.

And you're totally right about the SSMProblems integration. But it was convenient to recycle the structures.